### PR TITLE
Generate test specification from test cases

### LINKED ISSUE #439: Add framework for generating test spec
Nothing fancy, could be a README.md (markdown) in each test directory, or pydoc to extract the test's docstring in [reStructured Text](https://peps.python.org/pep-0287/) format.  

Either way, how tests are documented should be standardized: description, topology, pass criteria, allowed deviations, i.e., reasons for SKIP, e.g., tests running on virtual hardware.

The goal is to be able to extract the documentation for each test and generate a test specification based on the test suite.  Each test is numbered (e.g., section headings) and the test results should match those numbers in the test report.  The spec and report can be different documents.  Possibly everything is extracted to a `test-result/` directory with sub-directories for each test so any type of format can be generated.

### PR DESCRIPTION
**Pull request by mattiaswal**
## Description
This adds the possibility to generate the test specification.

More work remains, a lot of tests needs to be updated, to make it
possible for the test specification to be relevant for someone to
use.

This fixes #439 

## Checklist

Tick *relevant* boxes, this PR is-a or has-a:

- [ ] Bugfix
  - [ ] Regression tests
  - [ ] ChangeLog updates (for next release)
- [x] Feature
  - [ ] YANG model change => revision updated?
  - [ ] Regression tests added?
  - [ ] ChangeLog updates (for next release)
  - [x] Documentation added?
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (please detail in commit messages)
- [x] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

### CODE REVIEW

File: test/test.mk [hunk 1]
```diff
@@ -31,9 +33,13 @@ test:
 test-sh:
 	$(test-dir)/env $(base) $(mode) $(binaries) -i /bin/sh
 
+test-spec:
+	$(spec-dir)/generate_spec.py -d $(test-dir)/case
+	asciidoctor-pdf --theme $(spec-dir)/theme.yml -a pdf-fontsdir=$(spec-dir)/fonts -o $(test-specification) $(spec-dir)/Readme.adoc
```
Comment (troglobit) [inline test/test.mk:38]: Please add the [LICENSE.md](https://github.com/adobe-fonts/source-code-pro/blob/release/LICENSE.md) file to the fonts directory, they are free to use, but we still need to do our due diligence.

Please also mention the upstream reference/URL in the commit message that imports the fonts.  E.g., "Fonts imported from https://github.com/adobe-fonts/source-code-pro/"

**Review by troglobit — CHANGES_REQUESTED** [review]
Only a minor thing, otherwise it looks awesome! Nice work :clap: :clap: :clap:

**Review by troglobit — APPROVED** [review]
LGTM 👍

File: test/test.mk [hunk 2]
```diff
@@ -1,6 +1,8 @@
-base-dir := $(lastword $(subst :, ,$(BR2_EXTERNAL)))
-test-dir := $(BR2_EXTERNAL_INFIX_PATH)/test
-ninepm   := $(BR2_EXTERNAL_INFIX_PATH)/test/9pm/9pm.py
+base-dir           := $(lastword $(subst :, ,$(BR2_EXTERNAL)))
+test-dir           := $(BR2_EXTERNAL_INFIX_PATH)/test
+ninepm             := $(BR2_EXTERNAL_INFIX_PATH)/test/9pm/9pm.py
+spec-dir           := $(test-dir)/spec
+test-specification := $(spec-dir)/test-specification.pdf
```
Comment (wkz) [inline test/test.mk:5]: This will not be checked in, right? So should it live somewhere in output? `images/`?

File: test/spec/generate_spec.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            self.description=self.description.strip()
+        self.generic_visit(node)
+
+    # Check for test.step() for the actual test steps
+    def visit_Call(self, node):
+        if isinstance(node.func, ast.Attribute) and node.func.attr == 'step':
+            if isinstance(node.func.value, ast.Name) and node.func.value.id == 'test':
+                if node.args and isinstance(node.args[0], ast.Constant):
+                    self.test_steps.append(node.args[0].value)
+        self.generic_visit(node)  # Continue visiting other nodes
```
Comment (wkz) [inline test/spec/generate_spec.py:41]: So pretty :star_struck:

**Review by wkz — APPROVED** [review]
Love how this turned out! :heart:

**Pull request merged**